Add some simple sanity tests to ValueRanges

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94944
* #94939
* #94906
* __->__ #94905

To start, I simply test that unary/binary ops agree with reference when
the ranges are singleton.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Add some simple sanity tests to ValueRanges

To start, I simply test that unary/binary ops agree with reference when
the ranges are singleton.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add some simple sanity tests to ValueRanges"


To start, I simply test that unary/binary ops agree with reference when
the ranges are singleton.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

lezcano reviewed (approved):
For the general case, we can do it via sampling in the given region or via a grid with arange.

Also, this is missing testing `where`, but that shouldn't be a big deal.

ezyang commented:
> Also, this is missing testing where, but that shouldn't be a big deal.

yeah, I didn't do the non unary/binary because I don't need them for my use case (symbolic shapes)

> For the general case, we can do it via sampling in the given region or via a grid with arange.

It wasn't clear to me we would get much value from exhaustive fuzzing as opposed to well chosen constants, but I'm willing to change my mind

ezyang committed:
Update on "Add some simple sanity tests to ValueRanges"


To start, I simply test that unary/binary ops agree with reference when
the ranges are singleton.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

eellison commented on test/test_value_ranges.py:
```diff
@@ -0,0 +1,163 @@
+
+    @staticmethod
+    def exp(x):
+        return sympy.exp(x)
```
why use sympy instead of math ?

ezyang replied:
Right now x might not be a sympy expression, and I need it to get sympified. This will become less important once I make the invariant that internally it's only sympy.Expr.

eellison commented on test/test_value_ranges.py:
```diff
@@ -0,0 +1,163 @@
+    @parametrize("fn", BINARY_OPS)
+    def test_binary_ref(self, fn):
+        for a, b in itertools.product(CONSTANTS, repeat=2):
+            if fn == "pow" and (b > 4 or b == -1 or (a == b == 0)):
```
why the skips here ?

ezyang replied:
> 4 to avoid computing an extremely huge pow (which sympy will do symbolically), -1 is too strong I think I just need to prevent `-1 ** -1`, and the last cond is for `0 ** 0` which is undefined.

eellison reviewed (approved):
nice

ezyang committed:
Update on "Add some simple sanity tests to ValueRanges"


To start, I simply test that unary/binary ops agree with reference when
the ranges are singleton.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge
